rust-keybroker/keybroker-server/src/main.rs

```diff
@@ -1,6 +1,186 @@
 
-fn main() {
-    println!("Hello, world!");
+use std::sync::Mutex;
+
+use actix_web::{http, post, rt::task, web, App, HttpRequest, HttpResponse, HttpServer, Responder};
+use base64::engine::general_purpose::URL_SAFE_NO_PAD;
+use base64::prelude::*;
+use challenge::Challenger;
+use clap::Parser;
+use keybroker_common::{AttestationChallenge, BackgroundCheckKeyRequest, ErrorInformation};
+use keystore::KeyStore;
+mod challenge;
+mod error;
+mod keystore;
+mod verifier;
+
+#[post("/key/{keyid}")]
+async fn request_key(
+    path: web::Path<String>,
+    data: web::Data<ServerState>,
+    key_request: web::Json<BackgroundCheckKeyRequest>,
+) -> impl Responder {
+    let key_id = path.into_inner();
+
+    // Get a new challenge from the challenger.
+    let mut challenger = data.challenger.lock().expect("Poisoned challenger lock.");
+    let challenge = challenger.create_challenge(&key_id, &key_request.pubkey);
+
+    // TODO: The "accept" list is being hardcoded for Arm CCA here - it should come from the verifier.
+    let attestation_challenge = AttestationChallenge {
+        challenge: URL_SAFE_NO_PAD.encode(&challenge.challenge_value),
+        accept: vec![
+            "application/eat-collection; profile=http://arm.com/CCA-SSD/1.0.0".to_string(),
+        ],
+    };
+
+    let location = format!(
+        "{}:{}/keys/v1/evidence/{}",
+        data.args.baseurl, data.args.port, challenge.challenge_id
+    );
+
+    // TODO: Remove this println - tracer message for dev purposes only.
+    println!("Created attestation challenge at {}", location);
+
+    HttpResponse::Created()
+        .append_header((http::header::LOCATION, location))
+        .json(attestation_challenge)
+}
+
+#[post("/evidence/{challengeid}")]
+async fn submit_evidence(
+    path: web::Path<u32>,
+    data: web::Data<ServerState>,
+    request: HttpRequest,
+    evidence_base64: String,
+) -> impl Responder {
+    let challenge_id = path.into_inner();
+    let default_content_type = http::header::HeaderValue::from_static("application/string");
+
+    let challenge = {
+        let mut challenger = data.challenger.lock().expect("Poisoned challenger lock.");
+        let challenge = challenger.get_challenge(challenge_id);
+
+        if challenge.is_err() {
+            let error_info = ErrorInformation {
+                r#type: "AttestationFailure".to_string(),
+                detail: "The challenge identifier did not match any issued challenge.".to_string(),
+            };
+
+            return HttpResponse::Forbidden().json(error_info);
+        }
+
+        // Once the evidence is submitted, delete the challenge. It can't be used again.
+        challenger.delete_challenge(challenge_id).unwrap();
+
+        // This unwrap is now safe because we did the error check above.
+        challenge.unwrap()
+    };
+
+    // TODO: We are currently ignoring the content type from the request and assuming a CCA eat-collection.
+    let _content_type = request
```

Comment:
I've put together a quick test on the issue you were seeing when parsing the media type parameter.

The `profile` parameter value needs to be quoted, as in:
```rust
use mime;

fn main() {
    let mt = r#"application/eat-collection; profile="http://arm.com/CCA-SSD/1.0.0""#
        .parse::<mime::Mime>()
        .unwrap();
   assert_eq!(mt.get_param("profile").unwrap(), "http://arm.com/CCA-SSD/1.0.0");
}
```
The `mime` code chokes on `:` and `/` otherwise, because they are not valid [`token`](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6) elements.

Reply:
Thank you, yes, I was also prodding and discovered exactly the same, so good to have that validated independently. It took me a while to locate the `mime` crate, because I'd missed the clever little "trick" that Rust uses with `str::parse()` to invoke a trait on the target object through type inference. I need to do a quick re-test now to see whether the end-to-end circuit still works if I do the quoting in this way. If it does, I'll tweak this before we merge the PR.

Reply:
There might yet be a twist in this story. If I pass the quoted version of `Content-Type` to the key broker, the key broker accepts it, which solves the first half of the problem. However, if this string is then passed verbatim through to Veraison, it causes the attestation to fail. I need to see whether this is a problem in the Veraison Rust client or whether it's a server-side problem.

Reply:
Thanks for checking.  It's a bug in the Veraison plugin, see: https://github.com/veraison/services/blob/main/scheme/arm-cca/scheme.go
I'll raise a bug for that.

Reply:
https://github.com/veraison/services/issues/257